mods/lord/lord_money/shop.lua

```diff
@@ -152,180 +179,134 @@ minetest.register_craft({
+						end
+						minetest.log(
+							"action",
+							"магазин " .. pos .. " - игрок " .. name .. " пытался совершить обмен, но на складе не оказалось места"
+						)
+						minetest.chat_send_player(name, S("Exchange can not be done, ended place."))
+						return
+					end
 				end
-				pinv:add_item("customer_gets", stack) -- добавляем к игроку
 			end
-			-- пишем отчёт "всё в порядке" и выходим из функции.
-			minetest.log("action", "магазин "..minetest.pos_to_string(pos).." - игрок "..name.." произвёл обмен - всё в порядке")
-			minetest.chat_send_player(name, SL("Exchanged!"))
-		elseif fields.quit then -- выход с формы
-			local inv = sender:get_inventory()
-			for i = 1, inv:get_size("customer_gives") do -- для всех стеков в customer_gives
-				local stack = inv:get_stack("customer_gives", i)
-				if inv:room_for_item("main", stack) then inv:add_item("main", stack) -- если помещается, кидаем в main,
-				else minetest.item_drop(stack, sender, sender:get_pos()) -- если нет - кидаем на пол
+
+			-- ВРОДЕ ВСЁ НОРМАЛЬНО, ПРОИЗВОДИМ ОБМЕН
+				for _, stack in pairs(wants) do -- для всех стеков "Цена"
+					pinv:remove_item("customer_gives", stack) -- забираем у игрока из "Оплата"
+					if not is_endless then
+						minv:add_item("customers_gave", stack) -- помещаем в "Выручка"
+					end
 				end
-			end
-			inv:set_size("customer_gives", 0) -- удаляем customer_gives
-			for i = 1, inv:get_size("customer_gets") do -- для всех стеков в customer_gets
-				local stack = inv:get_stack("customer_gets", i)
-				if inv:room_for_item("main", stack) then inv:add_item("main", stack) -- если помещается, кидаем в main,
-				else minetest.item_drop(stack, sender, sender:get_pos()) -- если нет - кидаем на пол
+				for _, stack in pairs(gives) do -- для всех стеков "Предложение"
+					if not is_endless then
+						minv:remove_item("stock", stack) -- забираем со "Склад"
+					end
+					pinv:add_item("customer_gets", stack) -- добавляем игроку в "Приобретённый товар"
 				end
+				minetest.log(
+					"action",
+					"магазин " .. pos .. " - игрок " .. name .. " произвёл обмен - всё в порядке")
+				minetest.chat_send_player(name, S("Exchanged!"))
+
+
+		elseif fields.quit then -- выход с формы, возвращаем остатки игроку в инвентарь
+			for _,list in pairs({"customer_gives", "customer_gets"}) do
+			  if not pinv:is_empty(list) then
+			    for i, stack in ipairs(pinv:get_list(list)) do
+			      if pinv:room_for_item("main", stack) then
+			        pinv:add_item("main", stack) -- если помещается, кидаем в main,
+			      else
+			        minetest.log(
+			          "action",
+			          "магазин " .. pos .. " - игрок " .. name ..
+			          " инвентарь полон, товар " .. stack:get_name() .. " бросили рядом")
+			        minetest.item_drop(stack, sender, sender:get_pos()) -- если нет - кидаем на пол
+			      end
+			      pinv:set_stack(list, i, nil) -- Удаляем элемент i из list
+			    end
+			  end
 			end
```

Comment:
Тут отступы сломались

Reply:
Вот это поворот.... Исправил.
```suggestion
			for _,list in pairs({"customer_gives", "customer_gets"}) do
				if not pinv:is_empty(list) then
					for i, stack in ipairs(pinv:get_list(list)) do
						if pinv:room_for_item("main", stack) then
							pinv:add_item("main", stack) -- если помещается, кидаем в main,
						else
							minetest.log(
								"action",
								"магазин " .. pos .. " - игрок " .. name ..
								" инвентарь полон, товар " .. stack:get_name() .. " бросили рядом")
							minetest.item_drop(stack, sender, sender:get_pos()) -- если нет - кидаем на пол
						end
						pinv:set_stack(list, i, nil) -- Удаляем элемент i из list
					end
				end
			end
```